File: backend/app/api/projects.py
"""Project management API routes."""
from typing import List, Optional
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from app.api.auth import get_current_user, UserResponse

router = APIRouter()


# Models
class ProjectResponse(BaseModel):
    """Project response."""
    id: str
    name: str
    lyrics: Optional[str]
    music_genre: Optional[str]
    style: Optional[str]
    voice_id: Optional[str]
    music_url: Optional[str]
    vocal_url: Optional[str]
    final_url: Optional[str]
    status: str
    created_at: datetime


class ProjectCreate(BaseModel):
    """Project creation request."""
    name: str
    lyrics: Optional[str] = None
    music_genre: Optional[str] = None
    style: Optional[str] = None
    voice_id: Optional[str] = None


class GenerateRequest(BaseModel):
    """AI generation request."""
    prompt: str
    duration: Optional[int] = 180  # seconds


class ProjectUpdate(BaseModel):
    """Project update request."""
    name: Optional[str] = None
    lyrics: Optional[str] = None
    music_genre: Optional[str] = None
    style: Optional[str] = None
    voice_id: Optional[str] = None


# In-memory storage for mock data
PROJECTS_DB = {}


# Routes
@router.get("", response_model=List[ProjectResponse])
async def get_projects(current_user: UserResponse = Depends(get_current_user)):
    """Get user's projects."""
    # Return from mock DB
    user_projects = [p for p in PROJECTS_DB.values() if p["user_id"] == current_user.id]
    return user_projects


@router.post("", response_model=ProjectResponse)
async def create_project(
    project: ProjectCreate,
    current_user: UserResponse = Depends(get_current_user),
):
    """Create a new project."""
    project_id = f"proj-{len(PROJECTS_DB) + 1}"
    new_project = ProjectResponse(
        id=project_id,
        name=project.name,
        lyrics=project.lyrics,
        music_genre=project.music_genre,
        style=project.style,
        voice_id=project.voice_id,
        music_url=None,
        vocal_url=None,
        final_url=None,
        status="draft",
        created_at=datetime.utcnow(),
    )
    PROJECTS_DB[project_id] = {**new_project.model_dump(), "user_id": current_user.id}
    return new_project


@router.get("/{project_id}", response_model=ProjectResponse)
async def get_project(
    project_id: str,
    current_user: UserResponse = Depends(get_current_user),
):
    """Get project details."""
    if project_id not in PROJECTS_DB:
        raise HTTPException(status_code=404, detail="Project not found")
    return PROJECTS_DB[project_id]


@router.put("/{project_id}", response_model=ProjectResponse)
async def update_project(
    project_id: str,
    project: ProjectUpdate,
    current_user: UserResponse = Depends(get_current_user),
):
    """Update a project."""
    if project_id not in PROJECTS_DB:
        raise HTTPException(status_code=404, detail="Project not found")
    
    existing = PROJECTS_DB[project_id]
    if project.name is not None:
        existing["name"] = project.name
    if project.lyrics is not None:
        existing["lyrics"] = project.lyrics
    if project.music_genre is not None:
        existing["music_genre"] = project.music_genre
    if project.style is not None:
        existing["style"] = project.style
    if project.voice_id is not None:
        existing["voice_id"] = project.voice_id
    
    return existing


@router.post("/{project_id}/generate")
async def generate_song(
    project_id: str,
    request: GenerateRequest,
    current_user: UserResponse = Depends(get_current_user),
):
    """Trigger AI song generation."""
    # TODO: Call Suno + Reecho APIs
    return {
        "status": "processing",
        "message": "Generation started",
        "project_id": project_id,
    }


@router.delete("/{project_id}")
async def delete_project(
    project_id: str,
    current_user: UserResponse = Depends(get_current_user),
):
    """Delete a project."""
    if project_id not in PROJECTS_DB:
        raise HTTPException(status_code=404, detail="Project not found")
    
    # Verify ownership
    project = PROJECTS_DB[project_id]
    if project["user_id"] != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized to delete this project")
    
    del PROJECTS_DB[project_id]
    return {"status": "deleted", "project_id": project_id}

File: frontend/src/api/projects.ts
// Projects API client
import { getToken } from './auth';

const API_BASE = '/api/projects';

export interface Project {
  id: string;
  name: string;
  gpt_description?: string;
  lyrics?: string;
  music_genre?: string;
  style?: string;
  voice_id?: string;
  task_id?: string;
  suno_title?: string;
  suno_id?: string;
  mv: string;
  duration?: number;
  music_url?: string;
  cover_url?: string;
  vocal_url?: string;
  final_url?: string;
  status: string;
  error_message?: string;
  points_cost: number;
  created_at: string;
}

export interface ProjectCreate {
  name: string;
  gpt_description?: string;
  lyrics?: string;
  music_genre?: string;
  style?: string;
  voice_id?: string;
  mv?: string;
  make_instrumental?: boolean;
}

export interface TaskStatus {
  task_id: string;
  status: 'pending' | 'completed' | 'failed';
  audio_url?: string;
  cover_url?: string;
  duration?: number;
  title?: string;
  lyrics?: string;
  error_msg?: string;
  points_cost: number;
}

// Create a new project
export async function createProject(data: ProjectCreate): Promise<Project> {
  const token = getToken();
  const res = await fetch(API_BASE, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${token}`,
    },
    body: JSON.stringify(data),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: '创建失败' }));
    throw new Error(err.detail || '创建失败');
  }
  return res.json();
}

// Generate music for a project
export async function generateMusic(projectId: string): Promise<{ task_ids: number[]; message: string }> {
  const token = getToken();
  const res = await fetch(`${API_BASE}/${projectId}/generate`, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${token}`,
    },
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: '生成失败' }));
    throw new Error(err.detail || '生成失败');
  }
  return res.json();
}

// Check generation status
export async function checkStatus(projectId: string): Promise<TaskStatus> {
  const token = getToken();
  const res = await fetch(`${API_BASE}/${projectId}/status`, {
    headers: {
      'Authorization': `Bearer ${token}`,
    },
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: '查询失败' }));
    throw new Error(err.detail || '查询失败');
  }
  return res.json();
}

// Get all projects
export async function listProjects(): Promise<Project[]> {
  const token = getToken();
  const res = await fetch(API_BASE, {
    headers: {
      'Authorization': `Bearer ${token}`,
    },
  });
  if (!res.ok) {
    throw new Error('获取项目列表失败');
  }
  return res.json();
}

// Get single project
export async function getProject(projectId: string): Promise<Project> {
  const token = getToken();
  const res = await fetch(`${API_BASE}/${projectId}`, {
    headers: {
      'Authorization': `Bearer ${token}`,
    },
  });
  if (!res.ok) {
    throw new Error('获取项目失败');
  }
  return res.json();
}

// Delete project
export async function deleteProject(projectId: string): Promise<void> {
  const token = getToken();
  const res = await fetch(`${API_BASE}/${projectId}`, {
    method: 'DELETE',
    headers: {
      'Authorization': `Bearer ${token}`,
    },
  });
  if (!res.ok) {
    throw new Error('删除失败');
  }
}